### DISCUSSION TITLE
Access loaders data from different clients using `fetch`

### DISCUSSION THREAD

**Discussion by spacepolice10** [remix-run/remix#9354] [Q&A]
Hi there,

Is it possible to access JSON data through `await fetch` request made from client-side component or from different application? Right now if I try to fetch loader (like `fetch(**ROUTE_URL**).then(data => data.json())`) by its URL I receive the whole page (HTML) which is, I guess, correct but I would like to access JSON returned from loader via fetch. That will simplify making React Native version of application by exploiting Remix version as a backend. Any way to achieve something like that?

**Comment by sergiodxa** [comment]
You can use a resource route for that https://remix.run/docs/en/main/guides/resource-routes.

For UI routes you will need to use an internal ?_data=routeId search params or if you use single fetch .data?_route=routeId. Since this is already going to change in single fetch and may change again in the future without being a breaking change (it’s an internal implementation detail), the recommended way is to use a resource route.

**Reply by spacepolice10** [reply]
Ok, it means that there's no way to preserve Remix structure? Am I correct that I'll have to make a separate route for every possible request, then make separate UI routes and re-export resource routes there (or even use `fetcher.load`) to work it in Remix?

**Reply by kiliman** [reply]
I created a request filter that sits in front of the Remix request handler.

It checks the [`Sec-Fetch-Dest`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-Fetch-Dest) header to determine whether the request is a `document` or `fetch` one. 

It then uses the `matchRoutes` function to find the route `loader` and calls it to get the response.

So here's what it looks like when you do a `fetch('/counter')` request.

<img width="821" alt="image" src="https://github.com/remix-run/remix/assets/47168/42a9818e-5f79-4aeb-88e4-4582f5d18519">

Notice the `Sec-Fetch-Dest` header is `empty` so we know it comes from `fetch`.

<img width="750" alt="image" src="https://github.com/remix-run/remix/assets/47168/da019516-e2b7-480f-99ec-eb1f226d1256">

And here's the result: the actual JSON payload.

<img width="430" alt="image" src="https://github.com/remix-run/remix/assets/47168/b291a3e6-3379-41df-8935-5fd9672c339d">

**Reply by spacepolice10** [reply]
That sounds extremely interesting. Thanks for sharing & sorry to bother y'all again, but I feel like I don't really understand the conception. Do you read header in your **express** application, match route there & call certain loader (resource loader) right there (if so, how is it possible to activate loader is such a way before RemixRouter is in charge?)

**Reply by kiliman** [reply]
It'll be part of my Vite+Express plugin. It bundles your Express app with Remix so they have access to the same modules. It's also how I support Remix Middleware.

https://github.com/kiliman/remix-express-vite-plugin

**Reply by spacepolice10** [reply]
You mean that in future your package will by default intercept routes and return correct data depending on whether the request made by fetch or document?

**Reply by kiliman** [reply]
Yes, but it'll most likely be an opt-in since it would change the existing semantics.

**Reply by spacepolice10** [reply]
Sounds awesome. Starred the project, so I know when it is released. Look forward to try out this feature. Thanks
